File: word_pattern.py
# TIme : O(n) due to split else O(1)
# Space: O(n)
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        dict_letter = {}
        dict_words = {}
        s_arr = s.split(" ") # time: O(n), space: O(n)
        if len(pattern) != len(s_arr):
            return False
        for i in range(0, len(pattern)): # time O(n)
            if pattern[i] in dict_letter: # time O(1)
                if dict_letter.get(pattern[i]) != s_arr[i]: #time O(1)
                    return False
            else:
                dict_letter[pattern[i]] = s_arr[i] # time O(1), space: O(26 char * k) k: longest word
            if s_arr[i] in dict_words:
                if dict_words.get(s_arr[i]) != pattern[i]:
                    return False
            else:
                dict_words[s_arr[i]] = pattern[i] # time O(1), space: O(26 chars)
        return True

File: word_pattern.cpp
// Time Complexity : O(N+M)
// Space Complexity : O(N+26) : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : N/A

class Solution {
public:
    vector<string> splitStringToWords(string& s) // O(N), N = length of string s
    {
        std::vector<string> words;

        // create string stream from s 
        std::istringstream stream(s);
        std::string word;
    
        // Extract words separated by whitespace
        while (stream >> word) {
            words.push_back(word);
        }

        return words;
    }

    bool wordPattern(string pattern, string s) {

        if(pattern.length() == 0)
        {
            return false;
        }

        std::vector<string> words = splitStringToWords(s); // O(N)

        if(pattern.length() != words.size())
        {
            return false;
        }

        std::unordered_map<char, string> char_word_map;
        std::unordered_map<string, char> word_char_map;

        for(int i = 0; i < pattern.length(); i++) // (O(M))
        {
            char c = pattern[i];
            string word = words[i];

            if(char_word_map.contains(c))
            {
                if(char_word_map[c] != word)
                {
                    return false;
                }
            }
            else
            {
                char_word_map[c] = word;
            }


            if(word_char_map.contains(word))
            {
                if(word_char_map[word] != c)
                {
                    return false;
                }
            }
            else
            {
                word_char_map[word] = c;
            }
        }

        return true;
    }
};